providers/unleash/src/main/java/dev/openfeature/contrib/providers/unleash/UnleashProvider.java

```diff
@@ -0,0 +1,171 @@
+     */
+    @Override
+    public void initialize(EvaluationContext evaluationContext) throws Exception {
+        boolean initialized = isInitialized.getAndSet(true);
+        if (initialized) {
+            throw new GeneralError("already initialized");
+        }
+        super.initialize(evaluationContext);
+        UnleashSubscriberWrapper unleashSubscriberWrapper = new UnleashSubscriberWrapper(
+            unleashOptions.getUnleashConfigBuilder().build().getSubscriber(), this);
+        unleashOptions.getUnleashConfigBuilder().subscriber(unleashSubscriberWrapper);
+        UnleashConfig unleashConfig = unleashOptions.getUnleashConfigBuilder().build();
+        unleash = new DefaultUnleash(unleashConfig);
+
+        // else, state will be changed via UnleashSubscriberWrapper events
+        if (unleashConfig.isSynchronousFetchOnInitialisation()) {
+            state = ProviderState.READY;
+        } else {
+            log.info("ready state will be changed via UnleashSubscriberWrapper events");
+        }
+
+        log.info("finished initializing provider, state: {}", state);
+    }
+
+    @Override
+    public Metadata getMetadata() {
+        return () -> NAME;
+    }
+
+    @Override
+    public void emitProviderReady(ProviderEventDetails details) {
+        super.emitProviderReady(details);
+        state = ProviderState.READY;
+    }
+
+    @Override
+    public void emitProviderError(ProviderEventDetails details) {
+        super.emitProviderError(details);
+        state = ProviderState.ERROR;
+    }
+
+    @Override
+    public ProviderEvaluation<Boolean> getBooleanEvaluation(String key, Boolean defaultValue, EvaluationContext ctx) {
+        if (!ProviderState.READY.equals(state)) {
+            if (ProviderState.NOT_READY.equals(state)) {
+                throw new ProviderNotReadyError(PROVIDER_NOT_YET_INITIALIZED);
+            }
+            throw new GeneralError(UNKNOWN_ERROR);
+        }
+        UnleashContext context = ctx == null ? UnleashContext.builder().build() : ContextTransformer.transform(ctx);
+        boolean featureBooleanValue = unleash.isEnabled(key, context, defaultValue);
+        return ProviderEvaluation.<Boolean>builder()
+            .value(featureBooleanValue)
+            .build();
+    }
+
+    @Override
+    public ProviderEvaluation<String> getStringEvaluation(String key, String defaultValue, EvaluationContext ctx) {
+        if (!ProviderState.READY.equals(state)) {
+            if (ProviderState.NOT_READY.equals(state)) {
+                throw new ProviderNotReadyError(PROVIDER_NOT_YET_INITIALIZED);
+            }
+            throw new GeneralError(UNKNOWN_ERROR);
+        }
+        UnleashContext context = ctx == null ? UnleashContext.builder().build() : ContextTransformer.transform(ctx);
+        Variant evaluatedVariant = unleash.getVariant(key, context);
+        String variantName;
+        String value;
+        if (DISABLED_VARIANT.equals(evaluatedVariant)) {
```

Comment:
I am a bit unsure about this one. Here you are redefining the disabled variant in Unleash. 

Unleash itself has a concept of a "default" `Variant` you can pass and have that one returned by the SDK. This wrapper takes away the variant name and only keeps the value for the variant. Not sure if this is even possible to represent in Open Feature? It looks like a limitation in the API to only be able to define the value and not the name,

Reply:
**This is an important decision here.**

```
unleashProvider.getStringEvaluation("non-existing", "fallback_str", context);
```

What is the expected value here ?

- If we just return what the unleash API gives you, we will get **_null_**, as it returns DISABLED variant with a payload with null value.
- If we use implementation which checks for the DISABLED variant and uses default value, we will get **_fallback_str_**.

If first option is chosen, user can use default value with expecting it to be used in case of some issue like non-existing flag, but will actually get _null_.

@ivarconr @toddbaert what do you think ?

Reply:
Ah, I didn't realize the return value would be `null` in that case.

Then I think this implementation is the most idiomatic: 

> If we use implementation which checks for the DISABLED variant and uses default value, we will get fallback_str

You could also throw a FlagNotFoundError in this case, in which case the SDK would default. That would be more expensive because of the error serialization, but for this provider you might consider trying to get the variant in this DISABLED state an exceptional circumstance.

Reply:
When the user actually provide a defaultValue I think the current implementaion is ok for now, as long as we document that the variantName will not be set when the defaultValue is returned. 


But when the use **does not** provide a "defaultValue" it would be better to simply use the "DISBALED_VARIANT" from Unleash.